ros/angel_system_nodes/angel_system_nodes/object_descriptor_generator.py

```diff
@@ -0,0 +1,215 @@
+from cv_bridge import CvBridge
+import numpy as np
+import rclpy
+from rclpy.node import Node
+from sensor_msgs.msg import Image
+import torch
+import torchvision
+from torch.hub import load_state_dict_from_url
+from torchvision.models.detection.faster_rcnn import (
+    FasterRCNN,
+    fasterrcnn_resnet50_fpn,
+    model_urls
+)
+from torchvision.models.detection._utils import (
+    overwrite_eps
+)
+from torchvision.models.detection.backbone_utils import (
+    _validate_trainable_layers,
+    resnet_fpn_backbone
+)
+from torchvision import transforms
+
+from angel_msgs.msg import ObjectDescriptors
+
+
+BRIDGE = CvBridge()
+
+
+class ResNetDescriptors(FasterRCNN):
+    """
+    Custom ResNet class to override the forward method call
+    of the torch `FasterRCNN` class.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Creates a model with a `resnet_fpn_backbone`.
+
+        Copied from the torchvision.models.detection.faster_rcnn
+        faster_rcnn_resnet50_fpn function.
+        """
+        pretrained = True
+        trainable_backbone_layers = _validate_trainable_layers(
+            pretrained, None, 5, 3
+        )
+
+        backbone = resnet_fpn_backbone(
+            'resnet50',
+            False,
+            trainable_layers=trainable_backbone_layers
+        )
+
+        super().__init__(backbone=backbone, num_classes=91)
+
+        state_dict = load_state_dict_from_url(model_urls['fasterrcnn_resnet50_fpn_coco'],
+                                              progress=True)
+        self.load_state_dict(state_dict)
+        overwrite_eps(self, 0.0)
+
+    def forward(self, images, targets=None):
+        """
+        Overrides the FasterRCNN forward call to return the output of the
+        model backbone.
+
+        Modified version of thetorchvision `GeneralizedRCNN` forward call().
+        """
+        original_image_sizes: List[Tuple[int, int]] = []
+        for img in images:
+            val = img.shape[-2:]
+            torch._assert(
+                len(val) == 2,
+                f"expecting the last two dimensions of the Tensor to be H and W instead got {img.shape[-2:]}",
+            )
+            original_image_sizes.append((val[0], val[1]))
+
+        images, targets = self.transform(images, targets)
+
+        features = self.backbone(images.tensors)
+        return features
```

Comment:
I would want to get feedback from @cfunk1210 and/or Sri about what descriptors we want from the object detector. E.g. do we instead want a descriptor per proposal/detection, as opposed to this set of generic backbone features?